File: backend/app/schemas/document.py
"""
KnowledgeHub — Document Schemas
================================
Pydantic response models for Document and Chunk endpoints.
"""

from datetime import datetime
from uuid import UUID
from pydantic import BaseModel, ConfigDict


class ChunkResponse(BaseModel):
    """Pydantic model representing a single document chunk."""
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    document_id: UUID
    chunk_index: int
    content: str
    page_number: int | None
    character_count: int
    created_at: datetime


class DocumentUploadResponse(BaseModel):
    """Response returned upon initiating / completing document upload."""
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    filename: str
    original_filename: str
    file_size: int
    processing_status: str
    page_count: int | None
    chunk_count: int
    created_at: datetime


class DocumentListItem(BaseModel):
    """Summary representation of a document for list views."""
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    filename: str
    original_filename: str
    file_type: str
    file_size: int
    processing_status: str
    error_message: str | None = None
    page_count: int | None = None
    chunk_count: int = 0
    created_at: datetime
    updated_at: datetime


class DocumentListResponse(BaseModel):
    """Paginated list response for documents."""
    total: int
    documents: list[DocumentListItem]


class DocumentDetailResponse(BaseModel):
    """Full document details including its list of chunks."""
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    filename: str
    original_filename: str
    file_type: str
    file_size: int
    file_path: str
    source: str
    processing_status: str
    error_message: str | None = None
    page_count: int | None = None
    chunk_count: int = 0
    created_at: datetime
    updated_at: datetime
    chunks: list[ChunkResponse] = []

File: frontend/src/types/document.ts
export interface Chunk {
  id: string
  document_id: string
  chunk_index: number
  content: string
  page_number: number | null
  character_count: number
  created_at: string
}

export interface DocumentItem {
  id: string
  filename: string
  original_filename: string
  file_type: string
  file_size: number
  processing_status: 'pending' | 'processing' | 'completed' | 'failed'
  error_message?: string | null
  page_count?: number | null
  chunk_count: number
  created_at: string
  updated_at: string
}

export interface DocumentListResponse {
  total: number
  documents: DocumentItem[]
}

export interface DocumentDetail extends DocumentItem {
  file_path: string
  source: string
  chunks: Chunk[]
}

export interface DocumentUploadResponse {
  id: string
  filename: string
  original_filename: string
  file_size: number
  processing_status: string
  page_count?: number | null
  chunk_count: number
  created_at: string
}
